aixtend/utils/file_utils.py

```diff
@@ -14,50 +14,167 @@
     """
     session = requests.Session()
-    retries = Retry(total=5,
-                    backoff_factor=0.1,
-                    status_forcelist=[ 500, 502, 503, 504 ])
-    session.mount('https://', HTTPAdapter(max_retries=retries))
+    retries = Retry(total=5, backoff_factor=0.1, status_forcelist=[500, 502, 503, 504])
+    session.mount("https://", HTTPAdapter(max_retries=retries))
     response = session.request(method=method.upper(), url=url, **params)
     return response
-    
+
+
+def download_data(url_link, local_filename=None):
+    if local_filename is None:
+        local_filename = url_link.split("/")[-1]
+    with requests.get(url_link, stream=True) as r:
+        r.raise_for_status()
+        with open(local_filename, "wb") as f:
+            for chunk in r.iter_content(chunk_size=8192):
+                # If you have chunk encoded response uncomment if
+                # and set chunk_size parameter to None.
+                # if chunk:
+                f.write(chunk)
+    return local_filename
+
+
+def upload_data(
+    file_name: Union[Text, Path], content_type: Text = "text/csv", content_encoding: Optional[Text] = None, nattempts: int = 0
+):
+    """Upload files to S3 with pre-signed URLs
+
+    Args:
+        file_name (Union[Text, Path]): local path of file to be uploaded
+        content_type (Text, optional): Type of content. Defaults to "text/csv".
+        content_encoding (Text, optional): Content encoding. Defaults to None.
+        nattempts (int, optional): Number of attempts for diminish the risk of exceptions. Defaults to 0.
+
+    Returns:
+        URL: s3 path
+    """
+    try:
+        # Get pre-signed URL
+        team_key = config.TEAM_API_KEY
+        url = config.TEMPFILE_UPLOAD_URL
+
+        headers = {"Authorization": "token " + team_key}
+
+        payload = {"contentType": content_type, "originalName": file_name}
+        r = _request_with_retry("post", url, headers=headers, data=payload)
+        response = r.json()
+        path = response["key"]
+        # Upload data
+        presigned_url = response["uploadUrl"]
+        headers = {"Content-Type": content_type}
+        if content_encoding is not None:
+            headers["Content-Encoding"] = content_encoding
+        payload = open(file_name, "rb").read()
+        r = _request_with_retry("put", presigned_url, headers=headers, data=payload)
+
+        if r.status_code != 200:
+            if nattempts == 0:
+                return upload_data(file_name, content_type, content_encoding, nattempts + 1)
+            else:
+                raise Exception("File Uploading Error: Failure on Uploading to S3.")
+        bucket_name = re.findall(r"https://(.*?).s3.amazonaws.com", presigned_url)[0]
+        s3_link = f"s3://{bucket_name}/{path}"
+        return s3_link
+    except Exception as e:
+        if nattempts == 0:
+            return upload_data(file_name, content_type, content_encoding, nattempts + 1)
```

Comment:
Why not write the condition as nattempts > 1 and set nattempts = 2 in the initialization. Would be clearer to the user of this function.